### ISSUE TITLE
another zig cc regression in 0.9.0

### ISSUE THREAD

**Issue by travisstaloch** [ziglang/zig#10512]
The following runs fine with clang and previous versions of zig.  Not sure when it stopped working but seems to have happened in the past month as i first noticed this from a CI failure around 3 weeks ago (approx 12/7/21).  Seems related to #10386.  I am on a linux debian system: Linux 5.6.10-antix.1-amd64-smp

```c++
// tmp.cpp
#include <unordered_map>
#include <string>
#include <stdio.h>

const static std::unordered_map<char, int> ranks = {
  {'0', 0}, {'1', 1}, {'2', 2}, {'3', 3},
};

int main(int argc, char const *argv[])
{ 
  printf("ranks.at('2'): %d\n", ranks.at('2'));
  return 0;
}
```

```console
$ zig run tmp.cpp -lc++
libc++abi: terminating with uncaught exception of type std::out_of_range: unordered_map::at: key not found
Aborted
$ zig version
0.10.0-dev.63+6b8e33d14
``` 

This works fine when targeting musl: 
```console
$ zig run tmp.cpp -lc++ -target native-native-musl
ranks.at('2'): 2
```

Moving the the map to a non-static local in main also makes it work (without musl):

```cpp
// tmp2.cpp: works fine 
#include <unordered_map>
#include <string>
#include <stdio.h>


int main(int argc, char const *argv[])
{ 
  const std::unordered_map<char, int> ranks = {
    {'0', 0}, {'1', 1}, {'2', 2}, {'3', 3},
  };
  printf("ranks.at('2'): %d\n", ranks.at('2'));
  return 0;
}
```

**Comment by sagehane** [comment]
I wasn't sure if this stems from the same issue, but it seemed relevant so here it is.
I was trying to compile the Hello world program here: https://www.tutorialspoint.com/cplusplus-hello-world-program.

Fails on latest Zig:
```sh-session
$ zig version
0.10.0-dev.237+ada8e1713
$ zig run src/main.cpp -lc++
Segmentation fault (core dumped)
```

Works with `-target x86_64-linux-musl`:
```sh-session
$ zig run src/main.cpp -lc++ -target x86_64-linux-musl
Hello, World!
```

Works on Zig 0.8.1:
```sh-session
$ zig version
0.8.1
$ zig run src/main.cpp -lc++
Hello, World!
```

---

In case this matters, the `ldd` outputs of the binaries:
Master:
```sh-session
$ ldd a.out
	linux-vdso.so.1 (0x00007ffd255a4000)
	libpthread.so.0 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libpthread.so.0 (0x00007f31c2374000)
	libc.so.6 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libc.so.6 (0x00007f31c21af000)
	libdl.so.2 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libdl.so.2 (0x00007f31c21aa000)
	/nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/ld-linux-x86-64.so.2 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib64/ld-linux-x86-64.so.2 (0x00007f31c2396000)
```

0.8.1:
```sh-session
$ ldd a.out
	linux-vdso.so.1 (0x00007ffd2c3fc000)
	libm.so.6 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libm.so.6 (0x00007f40e34bc000)
	libpthread.so.0 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libpthread.so.0 (0x00007f40e349c000)
	libc.so.6 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libc.so.6 (0x00007f40e32d7000)
	libdl.so.2 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libdl.so.2 (0x00007f40e32d2000)
	librt.so.1 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/librt.so.1 (0x00007f40e32c7000)
	/nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/ld-linux-x86-64.so.2 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib64/ld-linux-x86-64.so.2 (0x00007f40e35ff000)
	libutil.so.1 => /nix/store/wl60dr9p15rwf53gxz61ijgisc1zdjc7-glibc-2.33-59/lib/libutil.so.1 (0x00007f40e32c0000)
```

**Comment by andrewrk** [comment]
Fix landed in 1e481dfef5a2241bfc5460214dbd0daffe5063bc.
